wgpu-core/src/macros.rs

```diff
@@ -0,0 +1,314 @@
+#[allow(unused_macros)]
+macro_rules! backends {
+    // one let statement per backend
+    (
+        #[vulkan] let $vk_field:pat = $vk_expr:expr;
+        #[metal] let $mtl_field:pat = $mtl_expr:expr;
+        #[dx12] let $dx12_field:pat = $dx12_expr:expr;
+        #[dx11] let $dx11_field:pat = $dx11_expr:expr;
+    ) => {
+        #[cfg(any(
+            windows,
+            all(unix, not(any(target_os = "ios", target_os = "macos"))),
+            feature = "gfx-backend-vulkan",
+        ))]
+        let $vk_field = $vk_expr;
+
+        #[cfg(any(target_os = "ios", target_os = "macos"))]
+        let $mtl_field = $mtl_expr;
+
+        #[cfg(windows)]
+        let $dx12_field = $dx12_expr;
+
+        #[cfg(windows)]
+        let $dx11_field = $dx11_expr;
+    };
+
+    // one block statement per backend
+    (
+        #[vulkan] $vk_block:block
+        #[metal] $mtl_block:block
+        #[dx12] $dx12_block:block
+        #[dx11] $dx11_block:block
+    ) => {
+        #[cfg(any(
+            windows,
+            all(unix, not(any(target_os = "ios", target_os = "macos"))),
+            feature = "gfx-backend-vulkan",
+        ))]
+        $vk_block
+
+        #[cfg(any(target_os = "ios", target_os = "macos"))]
+        $mtl_block
+
+        #[cfg(windows)]
+        $dx12_block
+
+        #[cfg(windows)]
+        $dx11_block
+    };
+
+    // a struct constructor with one field per backend
+    (
+        $Struct:path {
+            #[vulkan] $vk_field:ident: $vk_expr:expr,
+            #[metal] $mtl_field:ident: $mtl_expr:expr,
+            #[dx12] $dx12_field:ident: $dx12_expr:expr,
+            #[dx11] $dx11_field:ident: $dx11_expr:expr,
+        }
+    ) => {{
+        $Struct {
+            #[cfg(any(
+                windows,
+                all(unix, not(any(target_os = "ios", target_os = "macos"))),
+                feature = "gfx-backend-vulkan",
+            ))]
+            $vk_field: $vk_expr,
+
+            #[cfg(any(target_os = "ios", target_os = "macos"))]
+            $mtl_field: $mtl_expr,
+
+            #[cfg(windows)]
+            $dx12_field: $dx12_expr,
+
+            #[cfg(windows)]
+            $dx11_field: $dx11_expr,
+        }
+    }};
+}
+
+#[allow(unused_macros)]
```

Comment:
why would it be unused? it's used internally

Reply:
Whoops, artifact from development